util/merkle/src/node.rs

```diff
@@ -0,0 +1,110 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use elastic_array::ElasticArray1024;
+use hashdb::DBValue;
+use primitives::H256;
+use rlp::*;
+
+use nibbleslice::NibbleSlice;
+
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Node<'a> {
+    Leaf(NibbleSlice<'a>, DBValue),
+
+    Branch(NibbleSlice<'a>, [Option<H256>; 16]),
+}
+
+impl<'a> Node<'a> {
+    /// Decode the `node_rlp` and return the Node.
+    pub fn decoded(node_rlp: &'a [u8]) -> Option<Self> {
+        let r = Rlp::new(node_rlp);
+        match r.prototype() {
+            // Empty node
+            Prototype::Data(0) => return None,
+            // leaf node - first is nibble and second is value
+            Prototype::List(2) => {
+                let slice = NibbleSlice::from_encoded(r.at(0).data());
+
+                return Some(Node::Leaf(slice, DBValue::from_slice(&r.at(1).data())))
+            }
+            // branch node - first is nibbles (or empty), the rest 16 are nodes.
+            Prototype::List(17) => {
+                let mut nodes = [None; 16];
+                for i in 0..16 {
+                    nodes[i] = if r.at(i + 1).is_empty() {
+                        None
+                    } else {
+                        Some(r.at(i + 1).as_val::<H256>())
+                    }
+                }
+
+                return Some(Node::Branch(NibbleSlice::from_encoded(r.at(0).data()), nodes))
+            }
+
+            // something went wrong.
+            _ => panic!("Rlp data is not valid."),
+        }
+    }
+
+    /// Encode the node into RLP.
+    pub fn encoded(node: Option<Self>, size: Option<usize>) -> ElasticArray1024<u8> {
+        match node {
+            Some(Node::Leaf(slice, value)) => {
+                let mut stream = RlpStream::new_list(2);
+                stream.append(&&*slice.encoded());
+                stream.append(&&*value);
+                stream.drain()
+            }
+            Some(Node::Branch(slice, nodes)) => {
+                let mut stream = RlpStream::new_list(17);
+
+                match size {
+                    None => {
+                        stream.append(&&*slice.encoded());
+
+                        for child in nodes.iter() {
+                            if let Some(hash) = child {
+                                stream.append(hash);
+                            } else {
+                                stream.append_empty_data();
+                            }
+                        }
+                        stream.drain()
+                    }
+                    Some(s) => {
+                        stream.append(&&*slice.encoded_leftmost(s));
```

Comment:
```
let path = match size {
    None => &&*slice.encoded(),
    Some(s) => &&*slice.encoded_leftmost(s),
};
stream.append(path);
...
```

Reply:
it gives me this error
error[E0597]: borrowed value does not live long enough
  --> util/merkle/src/node.rs:78:33
   |
78 |                     Some(s) => &&*slice.encoded_leftmost(s),
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^-
   |                                 |                         |
   |                                 |                         temporary value dropped here while still borrowed
   |                                 temporary value does not live long enough
...
90 |             }
   |             - temporary value needs to live until here